pol-core/pol/getitem.cpp

```diff
@@ -291,28 +292,50 @@ void GottenItem::undo( Mobile::Character* chr )
 
   if ( _source == GOTTEN_ITEM_TYPE::GOTTEN_ITEM_IN_CONTAINER )
   {
-    // First attempt to put it back in the original container.
+    // First attempt to place the item in the player's backpack.
     UContainer* container = nullptr;
-    auto* orig_obj = system_find_object( _cnt_serial );
-    if ( orig_obj && orig_obj->isa( UOBJ_CLASS::CLASS_CONTAINER ) )
-    {
-      container = static_cast<UContainer*>( orig_obj );
-      if ( !container->can_add( *_item ) ||
+    if( !_item->no_drop() ) {
+      container = chr->backpack();
+      if ( !container || !container->can_add( *_item ) ||
            !container->can_insert_add_item( chr, UContainer::MT_PLAYER, _item ) )
         container = nullptr;
+      if ( _item->orphan() )
+        return;
     }
-    if ( _item->orphan() )
-      return;
-    if ( !container )
+    // Attempt to put it back in the original container.
+    if ( !container
+         && ( !Core::settingsManager.ssopt.undo_get_item_drop_here || _item->no_drop() ) )
     {
-      // Attempt to place the item in the player's backpack.
+      auto* orig_obj = system_find_object( _cnt_serial );
+      if ( orig_obj && orig_obj->isa( UOBJ_CLASS::CLASS_CONTAINER ) )
+      {
+        if ( _item->no_drop()
+             || chr->can_moveanydist()
+             || !Core::settingsManager.ssopt.undo_get_item_enable_range_check
+             || chr->pos().pol_distance( _pos )
```

Comment:
As discussed in Discord, should check the distance to `orig_obj->toplevel_pos()` and not `_pos`, as this is the position inside the gump container.